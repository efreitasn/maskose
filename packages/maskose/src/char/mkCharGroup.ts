import { MaskCharPredicateFn, MaskChar } from '.';

export const MASK_CHAR_GROUP_TYPE = 'MASK_CHAR_GROUP_TYPE';

export type MaskCharGroup = {
  type: typeof MASK_CHAR_GROUP_TYPE;
  maskChars: MaskChar[];
  predicateFn?: MaskCharPredicateFn;
};

/**
 * A group of characters
 * @param maskChars The characters that will be in the group
 * @param predicateFn Function that will decide if the character will be present in the mask
 */
export default function mkCharGroup(
  maskChars: MaskChar[],
  predicateFn?: MaskCharPredicateFn
): MaskCharGroup {
  return {
    type: MASK_CHAR_GROUP_TYPE,
    maskChars,
    predicateFn
  };
}