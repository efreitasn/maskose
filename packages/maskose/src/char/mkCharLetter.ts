import { MaskCharPredicateFn } from '.';

export const MASK_CHAR_LETTER_TYPE = 'MASK_CHAR_LETTER_TYPE';

export type MaskCharLetterObj = {
  type: typeof MASK_CHAR_LETTER_TYPE;
  predicateFn?: MaskCharPredicateFn;
};

/**
 * A character in the range A to Z (case insensitive) expected to be in the value to be masked
 * @param predicateFn Function that will decide if the character will be present in the mask
 */
export default function mkCharLetter(predicateFn?: MaskCharPredicateFn): MaskCharLetterObj {
  return {
    type: MASK_CHAR_LETTER_TYPE,
    predicateFn
  };
}