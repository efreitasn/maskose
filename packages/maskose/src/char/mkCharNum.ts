import { MaskCharPredicateFn } from '.';

export const MASK_CHAR_NUM_TYPE = 'MASK_CHAR_NUM_TYPE';

export type MaskCharNumObj = {
  type: typeof MASK_CHAR_NUM_TYPE;
  predicateFn?: MaskCharPredicateFn;
};

/**
 * A character in the range 0 to 9 expected to be in the value to be masked
 * @param predicateFn Function that will decide if the character will be present in the mask
 */
export default function mkCharNum(predicateFn?: MaskCharPredicateFn): MaskCharNumObj {
  return {
    type: MASK_CHAR_NUM_TYPE,
    predicateFn
  };
}