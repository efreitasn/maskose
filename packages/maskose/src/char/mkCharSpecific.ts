import { MaskCharPredicateFn } from '.';

export const MASK_CHAR_SPECIFIC_TYPE = 'MASK_CHAR_SPECIFIC_TYPE';

export type MaskCharSpecificObj = {
  type: typeof MASK_CHAR_SPECIFIC_TYPE;
  char: string;
  predicateFn?: MaskCharPredicateFn;
};

/**
 * A specific character expected to be in the value to be masked
 * @param char Character that will be used in the mask
 * @param predicateFn Function that will decide if the character will be present in the mask
 */
export default function mkCharSpecific(char: string, predicateFn?: MaskCharPredicateFn): MaskCharSpecificObj {
  return {
    type: MASK_CHAR_SPECIFIC_TYPE,
    char,
    predicateFn
  };
}