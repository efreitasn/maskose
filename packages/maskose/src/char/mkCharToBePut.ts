import { MaskCharPredicateFn } from '.';

const MASK_CHAR_TO_BE_PUT_TYPE = 'MASK_CHAR_TO_BE_PUT_TYPE';

export type MaskCharToBePutObj = {
  type: typeof MASK_CHAR_TO_BE_PUT_TYPE;
  char: string;
  predicateFn?: MaskCharPredicateFn;
};

/**
 * A character that's expected to be put in the mask when formatting instead of being a character
 * expected to be in the value to be masked. For example, the dash (-) in a phone number mask is
 * such a character
 * @param char Character to be put in the mask
 * @param predicateFn Function that will decide if the character will be present in the mask
 */
export default function mkCharToBePut(char: string, predicateFn?: MaskCharPredicateFn): MaskCharToBePutObj {
  return {
    type: MASK_CHAR_TO_BE_PUT_TYPE,
    char,
    predicateFn
  };
}