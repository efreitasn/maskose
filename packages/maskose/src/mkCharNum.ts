const MASK_CHAR_NUM_TYPE = 'MASK_CHAR_NUM_TYPE';

export type MaskCharPredicateFn = (arg: {
  value: string;
}) => boolean;

export type MkCharNumObj = {
  type: typeof MASK_CHAR_NUM_TYPE;
  predicateFn?: MaskCharPredicateFn;
};

export default function mkCharNum(predicateFn?: MaskCharPredicateFn): MkCharNumObj {
  return {
    type: MASK_CHAR_NUM_TYPE,
    predicateFn
  };
}