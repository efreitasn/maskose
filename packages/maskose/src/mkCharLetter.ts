import { MaskCharPredicateFn } from './mkCharNum';

const MASK_CHAR_LETTER_TYPE = 'MASK_CHAR_LETTER_TYPE';

export type MkCharLetterObj = {
  type: typeof MASK_CHAR_LETTER_TYPE;
  predicateFn?: MaskCharPredicateFn;
};

export default function mkCharLetter(predicateFn?: MaskCharPredicateFn): MkCharLetterObj {
  return {
    type: MASK_CHAR_LETTER_TYPE,
    predicateFn
  };
}