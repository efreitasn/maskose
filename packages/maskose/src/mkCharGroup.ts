import { MaskCharPredicateFn, MkCharNumObj } from './mkCharNum';
import { MkCharLetterObj } from './mkCharLetter';
import { MkCharToBePutObj } from './mkCharToBePut';
import { MkCharSpecificObj } from './mkCharSpecific';

const MASK_CHAR_GROUP_TYPE = 'MASK_CHAR_GROUP_TYPE';

export type CharObj =
  | MkCharLetterObj
  | MkCharSpecificObj
  | MkCharNumObj
  | MkCharToBePutObj
;

export type MkCharGroupObj = {
  type: typeof MASK_CHAR_GROUP_TYPE;
  charObjs: CharObj[];
  predicateFn?: MaskCharPredicateFn;
};

/**
 * A group of characters
 * @param charObjs The characters that will be in the group
 * @param predicateFn Function that will decide if the character will be present in the mask
 */
export default function mkCharGroup(charObjs: CharObj[], predicateFn?: MaskCharPredicateFn): MkCharGroupObj {
  return {
    type: MASK_CHAR_GROUP_TYPE,
    charObjs,
    predicateFn
  };
}