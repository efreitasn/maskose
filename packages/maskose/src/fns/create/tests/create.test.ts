import mkCreate from '..';
import mkCharBoostRepeat from '../../../boosts/char/repeat';
import mkBoostChar from '../../boostChar';
import mkCharBoostValueLengthEqualTo from '../../../boosts/char/valueLengthEqualTo';
import mkCharNum from '../../../mask/chars/num';
import mkCharLetter from '../../../mask/chars/letter';
import mkCharGroup from '../../../mask/chars/group';
import mkCharToBePut from '../../../mask/chars/toBePut';
import mkCharSpecific from '../../../mask/chars/specific';

test('should match the object returned by mkCreate()', () => {
  const result = mkCreate([
    mkCharBoostRepeat(4)(mkCharNum()),
    mkCharLetter(),
    mkCharBoostRepeat(2)(
      mkCharGroup([
        mkCharNum(),
        mkCharLetter()
      ])
    ),
    mkBoostChar(mkCharToBePut(')'))([
      mkCharBoostValueLengthEqualTo(2)
    ]),
    mkCharToBePut('K'),
    mkCharNum(),
    mkCharSpecific('B'),
    mkCharGroup([
      mkCharNum(),
      mkCharLetter()
    ])
  ]);

  expect(result).toMatchSnapshot();
});