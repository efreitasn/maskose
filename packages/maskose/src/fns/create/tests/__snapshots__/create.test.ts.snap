// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the object returned by mkCreate() 1`] = `
Object {
  "chars": Array [
    Object {
      "maskedValueLengthConditions": Array [],
      "regExp": /\\[0-9\\]/,
      "repetitions": 4,
      "type": "MASKOSE_CHAR_NUM_TYPE",
      "valueToBeMaskedLengthConditions": Array [],
    },
    Object {
      "maskedValueLengthConditions": Array [],
      "regExp": /\\[A-Z\\]/i,
      "repetitions": 1,
      "type": "MASKOSE_CHAR_LETTER_TYPE",
      "valueToBeMaskedLengthConditions": Array [],
    },
    Object {
      "chars": Array [
        Object {
          "maskedValueLengthConditions": Array [],
          "regExp": /\\[0-9\\]/,
          "repetitions": 1,
          "type": "MASKOSE_CHAR_NUM_TYPE",
          "valueToBeMaskedLengthConditions": Array [],
        },
        Object {
          "maskedValueLengthConditions": Array [],
          "regExp": /\\[A-Z\\]/i,
          "repetitions": 1,
          "type": "MASKOSE_CHAR_LETTER_TYPE",
          "valueToBeMaskedLengthConditions": Array [],
        },
      ],
      "maskedValueLengthConditions": Array [],
      "repetitions": 2,
      "type": "MASKOSE_CHAR_GROUP_TYPE",
      "valueToBeMaskedLengthConditions": Array [],
    },
    Object {
      "char": ")",
      "maskedValueLengthConditions": Array [
        Object {
          "num": 2,
          "type": "eq",
        },
      ],
      "regExp": /\\\\\\)/,
      "repetitions": 1,
      "type": "MASKOSE_CHAR_TO_BE_PUT_TYPE",
      "valueToBeMaskedLengthConditions": Array [
        Object {
          "num": 2,
          "type": "eq",
        },
      ],
    },
    Object {
      "char": "K",
      "maskedValueLengthConditions": Array [],
      "regExp": /K/,
      "repetitions": 1,
      "type": "MASKOSE_CHAR_TO_BE_PUT_TYPE",
      "valueToBeMaskedLengthConditions": Array [],
    },
    Object {
      "maskedValueLengthConditions": Array [],
      "regExp": /\\[0-9\\]/,
      "repetitions": 1,
      "type": "MASKOSE_CHAR_NUM_TYPE",
      "valueToBeMaskedLengthConditions": Array [],
    },
    Object {
      "char": "B",
      "maskedValueLengthConditions": Array [],
      "regExp": /B/,
      "repetitions": 1,
      "type": "MASKOSE_CHAR_SPECIFIC_TYPE",
      "valueToBeMaskedLengthConditions": Array [],
    },
    Object {
      "chars": Array [
        Object {
          "maskedValueLengthConditions": Array [],
          "regExp": /\\[0-9\\]/,
          "repetitions": 1,
          "type": "MASKOSE_CHAR_NUM_TYPE",
          "valueToBeMaskedLengthConditions": Array [],
        },
        Object {
          "maskedValueLengthConditions": Array [],
          "regExp": /\\[A-Z\\]/i,
          "repetitions": 1,
          "type": "MASKOSE_CHAR_LETTER_TYPE",
          "valueToBeMaskedLengthConditions": Array [],
        },
      ],
      "maskedValueLengthConditions": Array [],
      "repetitions": 1,
      "type": "MASKOSE_CHAR_GROUP_TYPE",
      "valueToBeMaskedLengthConditions": Array [],
    },
  ],
  "direction": "left-to-right",
  "endless": false,
}
`;
